src/Program2.cs

```diff
@@ -0,0 +1,526 @@
+﻿using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Sources;
+
+//class Doc
+//{
+//    async IAsyncEnumerable<int> Illustration()
+//    {
+//        // call to WaitForNext will start all the code until state 1
+//        Console.Write("abc");
+//        await Slow(); // state 0, machine automatically continues
+
+//        Console.Write("def");
+//        await Slow(); // state 2, machine automatically continues
+
+//        Console.Write("ghi");
+//        yield return 42; // state 1, promise of a value is fulfilled, WaitForNext completes with true, promise is left as signal of available value
+//                         // first call to TryGetNext notices promise from running machine, clears it, returns value and true, moves to state 3
+//                         // second call to TryGetNext prints "jkl", has no promise to fulfill, returns true, moves to state 3
+
+//        Console.Write("jkl");
+//        yield return 43; // state 3
+//                         // third call to TryGetNext prints "mno", has no promise to fulfill, returns false, moves to state 4, lets machine run asynchronously
+
+//        Console.Write("mno");
+//        await Slow(); // state 4, machine automatically continues
+//                      // call to WaitForNext in parallel with code executing already
+
+//        Console.Write("pqr");
+//        await Slow(); // state 6, machine automatically continues
+
+//        Console.Write("stu");
+//        yield return 44; // state 5, promise of a value is fulfilled, WaitForNext completes with true
+//                         // fourth call to TryGetNext returns true, moves to state -2
+//                         // fifth call to TryGetNext returns false
+//                         // call to WaitForNext returns false
+//    }
+
+//    Task Slow() => throw null;
+//}
+
+class Program
+{
+    static async Task Main()
+    {
+        IAsyncEnumerable<int> test = AsyncIterator();
+        IAsyncEnumerator<int> enumerable = test.GetAsyncEnumerator();
+
+        //foreach await (var value in enumerable)
+        //{
+        //    Console.WriteLine(value)
+        //}
+        while (await enumerable.WaitForNextAsync())
+        {
+            while (true)
+            {
+                int value = enumerable.TryGetNext(out bool success);
+                if (!success) goto outer_loop_continue;
+
+                Console.WriteLine(value);
+            }
+            outer_loop_continue:;
+        }
+    }
+
+    static IAsyncEnumerable<int> AsyncIterator()
+    {
+        var stateMachine = new Unprounouncable();
+        stateMachine.State = StateMachineStates.NotStartedStateMachine;
+        stateMachine.Builder = AsyncTaskMethodBuilder.Create();
```

Comment:
As long as we're talking about adding new types to the framework, we might want to introduce a specialized builder for this purpose, one that includes AwaitUnsafeOnCompleted and the equivalent of Start (if we want to handle ExecutionContext flow as async methods do), but that doesn't have the notion of completing a Task.  Functionally we can use AsyncTaskMethodBuilder, but it's going to incur an unnecessary allocation for each enumerable, because it allocates a Task that won't ever be completed or returned.  I used this in my prototype as well, though: https://github.com/dotnet/corefx/blob/ecf20bcd7ce610c6cd242441c44af4389295757d/src/System.Threading.Tasks.Extensions/tests/ManualResetValueTaskSourceTests.cs#L144